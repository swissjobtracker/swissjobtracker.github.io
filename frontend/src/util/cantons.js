export default [
  'ag',
  'ai',
  'ar',
  'be',
  'bl',
  'bs',
  'fr',
  'ge',
  'gl',
  'gr',
  'ju',
  'lu',
  'ne',
  'nw',
  'ow',
  'sg',
  'sh',
  'so',
  'sz',
  'tg',
  'ti',
  'ur',
  'vd',
  'vs',
  'zg',
  'zh'
].sort()
