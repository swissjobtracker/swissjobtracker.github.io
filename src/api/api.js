const throwStatusTextIfNotOk = (response) => {
  if (!response.ok) {
    throw new Error(response.statusText);
  }

  return response;
};

const getTimeseries = (keys) => {
  return (
    fetch(
      `https://datenservice.kof.ethz.ch/api/v1/public/ts?keys=${keys.join(
        ",",
      )}&df=Y-m-d`,
    )
      // TODO: Handle errors
      .then(throwStatusTextIfNotOk)
      .then((r) => r.json())
  );
};

export default {
  getTimeseries,
};
