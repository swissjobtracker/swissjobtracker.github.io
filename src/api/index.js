// TODO: conditional import of fake or real api based on env var
//       mayhaps need to conditionally EXport tho, or top level await here
import ts from './fakeapi'

export default {
  ...ts
}
