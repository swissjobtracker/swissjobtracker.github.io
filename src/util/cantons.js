export default [
  {
    label: 'Aargau',
    value: 'ag',
  },
  {
    label: 'Appenzell Innerrhoden',
    value: 'ai',
  },
  {
    label: 'Appenzell Ausserrhoden',
    value: 'ar',
  },
  {
    label: 'Bern',
    value: 'be',
  },
  {
    label: 'Basel-Landschaft',
    value: 'bl',
  },
  {
    label: 'Basel-Stadt',
    value: 'bs',
  },
  {
    label: 'Fribourg',
    value: 'fr',
  },
  {
    label: 'Genève',
    value: 'ge',
  },
  {
    label: 'Glarus',
    value: 'gl',
  },
  {
    label: 'Grishun',
    value: 'gr',
  },
  {
    label: 'Jura',
    value: 'ju',
  },
  {
    label: 'Luzern',
    value: 'lu',
  },
  {
    label: 'Neuchâtel',
    value: 'ne',
  },
  {
    label: 'Nidwalden',
    value: 'nw',
  },
  {
    label: 'Obwalden',
    value: 'ow',
  },
  {
    label: 'St. Gallen',
    value: 'sg',
  },
  {
    label: 'Schaffhausen',
    value: 'sh',
  },
  {
    label: 'Solothurn',
    value: 'so',
  },
  {
    label: 'Schwyz',
    value: 'sz',
  },
  {
    label: 'Thurgau',
    value: 'tg',
  },
  {
    label: 'Ticino',
    value: 'ti',
  },
  {
    label: 'Uri',
    value: 'ur',
  },
  {
    label: 'Vaud',
    value: 'vd',
  },
  {
    label: 'Vallais',
    value: 'vs',
  },
  {
    label: 'Zug',
    value: 'zg',
  },
  {
    label: 'Zürich',
    value: 'zh'
  }
].sort((a, b) => a.value > b.value ? 1 : -1)
