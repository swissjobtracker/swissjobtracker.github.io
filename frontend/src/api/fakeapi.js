import { formatDate } from '../util/dates'

const generators = [
  (i, max) => 10*(i/max),
  (i, max) => 10*Math.sin(4*Math.PI*(i/max)),
  (i, max) => 10*((2*i/max)*100 % 1),
  (i, max) => 10*Math.cos(4*Math.PI*(i/max)),
  (i, max) => 7,
  (i, max) => 10*(1-i/max),
  (i, max) => 10*Math.random(),
  (i, max) => 10*Math.sin(Math.PI*(i/max))
]

// spot the StackOverflow c&p ;P
function shuffleArray(array) {
  for (var i = array.length - 1; i > 0; i--) {
      var j = Math.floor(Math.random() * (i + 1));
      var temp = array[i];
      array[i] = array[j];
      array[j] = temp;
  }
}

export const getFakeData = (keys, n, seq = false) => {
  const d0 = new Date("2021-01-01")
  const dates = (new Array(n)).fill(0).map((x, i) => {
    let d = new Date(d0)
    d.setDate(d.getDate() + 7*i)
    return d
  })
  const genI = (new Array(keys.length)).fill(0).map((x, i) => i % generators.length)
  shuffleArray(genI)
  const gens = genI.map((x) => seq ? ((i, n) => i) : generators[x])
  return Object.fromEntries(keys.map((k, j) => [k, (new Array(n)).fill(0).map((x, i) => { return {date: formatDate(dates[i]), value: gens[j](i, n)}})]))
}

const getTimeseries = (keys) => new Promise((resolve, reject) => resolve(getFakeData(keys, 123)))

export default {
  getTimeseries
}
