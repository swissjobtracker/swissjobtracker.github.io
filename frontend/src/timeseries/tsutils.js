import cantons from '../util/cantons'
import { formatDate } from '../util/dates'

export const tsToMap = (data, t) => {
  const date = formatDate(t)
  const tsKeys = Object.keys(data).sort() // kinda assuming that everything before canton is constant...
  return cantons.map((c, i) => ({
    name: c.toUpperCase(),
    // TODO: Do better e.g. handle date not in array
    value: data[tsKeys[i]].filter((x) => x.date == date)[0].value
  }))
}

export const tsToLine = (data) => {
  const keys = Object.keys(data)
  const nSeries = keys.length
  let dates = new Set()
  for(const [key, value] of Object.entries(data)) {
    value.map(({ date }) => date).forEach(dates.add, dates)
  }

  let data_obj = {};
  dates.forEach((dat) => {
      data_obj[dat] = new Array(nSeries).fill(null);
  })

  for(const i in keys) {
    data[keys[i]].forEach(({ date, value }) => data_obj[date][i] = value)
  }

  return Object.entries(data_obj).sort((a, b) => a[0] > b[0] ? 1 : -1).map(([d, v]) => [d, ...v])
}

export default {
  tsToMap,
  tsToLine
}
