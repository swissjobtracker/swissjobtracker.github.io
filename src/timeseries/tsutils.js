import cantons from '../util/cantons'
import { formatDate } from '../util/dates'

/**
 * Transform api timeseries data to echarts map input at time t
 * @param {tsdata} data
 * @param {string} t
 * @returns [{ name: 'XX', value: yy }, ...] for all cantons
 */
export const tsToMap = (data, t) => {
  if(t === null) {
    const dates = Object.entries(data)
                        .map(([_, ts]) => (new Date(ts[ts.length - 1].date)))
    t = Math.max(...dates)
  }

  const date = formatDate(t)
  // keys are identical up to canton chunk so sorting by the whole key effectively means sorting by canton
  // e.g. ch.kof.jobtracker.canton...
  const tsKeys = Object.keys(data).sort()
  return cantons.map((c, i) => ({
    name: c.value.toUpperCase(),
    // TODO: Do better e.g. handle date not in array
    value: data[tsKeys[i]].filter((x) => x.date == date)[0].value
  }))
}

/**
 * Transforms api timeseries data to echarts line chart format
 * @param {tsdata} data
 * @returns
 */
export const tsToLine = (data) => {
  const keys = Object.keys(data)
  const nSeries = keys.length
  let dates = new Set()
  for(const [key, value] of Object.entries(data)) {
    value.map(({ date }) => date).forEach(dates.add, dates)
  }

  let data_obj = {};
  dates.forEach((dat) => {
      data_obj[dat] = new Array(nSeries).fill(null);
  })

  for(const i in keys) {
    data[keys[i]].forEach(({ date, value }) => data_obj[date][i] = value)
  }

  const values =  Object.entries(data_obj).sort((a, b) => a[0] > b[0] ? 1 : -1).map(([d, v]) => [d, ...v])
  return [['date', ...keys], ...values]
}

export default {
  tsToMap,
  tsToLine
}
