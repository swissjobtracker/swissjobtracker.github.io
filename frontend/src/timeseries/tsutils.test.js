import { getFakeData } from '../api/fakeapi'
import cantons from '../util/cantons'
import tsutils from './tsutils'

test('tsToMap', () => {
  expect(tsutils.tsToMap(getFakeData(cantons, 10, true), new Date('2021-01-01'))).toMatchSnapshot()
})

test('tsToLine', () => {
  expect(tsutils.tsToLine(getFakeData(cantons, 10, true))).toMatchSnapshot()
})
