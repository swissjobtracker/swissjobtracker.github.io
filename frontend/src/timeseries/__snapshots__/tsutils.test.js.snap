// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tsToLine 1`] = `
Array [
  Array [
    "date",
    "ag",
    "ai",
    "ar",
    "be",
    "bl",
    "bs",
    "fr",
    "ge",
    "gl",
    "gr",
    "ju",
    "lu",
    "ne",
    "nw",
    "ow",
    "sg",
    "sh",
    "so",
    "sz",
    "tg",
    "ti",
    "ur",
    "vd",
    "vs",
    "zg",
    "zh",
  ],
  Array [
    "2021-01-01",
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    "2021-01-08",
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
  ],
  Array [
    "2021-01-15",
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
    2,
  ],
  Array [
    "2021-01-22",
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
    3,
  ],
  Array [
    "2021-01-29",
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
    4,
  ],
  Array [
    "2021-02-05",
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
    5,
  ],
  Array [
    "2021-02-12",
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
    6,
  ],
  Array [
    "2021-02-19",
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
    7,
  ],
  Array [
    "2021-02-26",
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
    8,
  ],
  Array [
    "2021-03-05",
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
    9,
  ],
]
`;

exports[`tsToMap 1`] = `
Array [
  Object {
    "name": "AG",
    "value": 0,
  },
  Object {
    "name": "AI",
    "value": 0,
  },
  Object {
    "name": "AR",
    "value": 0,
  },
  Object {
    "name": "BE",
    "value": 0,
  },
  Object {
    "name": "BL",
    "value": 0,
  },
  Object {
    "name": "BS",
    "value": 0,
  },
  Object {
    "name": "FR",
    "value": 0,
  },
  Object {
    "name": "GE",
    "value": 0,
  },
  Object {
    "name": "GL",
    "value": 0,
  },
  Object {
    "name": "GR",
    "value": 0,
  },
  Object {
    "name": "JU",
    "value": 0,
  },
  Object {
    "name": "LU",
    "value": 0,
  },
  Object {
    "name": "NE",
    "value": 0,
  },
  Object {
    "name": "NW",
    "value": 0,
  },
  Object {
    "name": "OW",
    "value": 0,
  },
  Object {
    "name": "SG",
    "value": 0,
  },
  Object {
    "name": "SH",
    "value": 0,
  },
  Object {
    "name": "SO",
    "value": 0,
  },
  Object {
    "name": "SZ",
    "value": 0,
  },
  Object {
    "name": "TG",
    "value": 0,
  },
  Object {
    "name": "TI",
    "value": 0,
  },
  Object {
    "name": "UR",
    "value": 0,
  },
  Object {
    "name": "VD",
    "value": 0,
  },
  Object {
    "name": "VS",
    "value": 0,
  },
  Object {
    "name": "ZG",
    "value": 0,
  },
  Object {
    "name": "ZH",
    "value": 0,
  },
]
`;
